File: clients/ruby/lib/chromadb/embedding_functions/chroma_cloud_splade.rb
# frozen_string_literal: true

module Chroma
  module EmbeddingFunctions
    class ChromaCloudSpladeEmbeddingFunction
      NAME = "chroma-cloud-splade"
      DEFAULT_MODEL = "prithivida/Splade_PP_en_v1"

      attr_reader :api_key_env_var, :model, :include_tokens

      def initialize(api_key_env_var: "CHROMA_API_KEY", model: DEFAULT_MODEL, include_tokens: false)
        @api_key_env_var = api_key_env_var
        @api_key = ENV[@api_key_env_var] || Chroma::SharedState.cloud_api_key
        raise ArgumentError, "API key not found in #{@api_key_env_var} or any existing clients" if @api_key.nil? || @api_key.empty?

        @model = model
        @include_tokens = !!include_tokens

        @connection = Faraday.new(url: "https://embed.trychroma.com") do |builder|
          builder.headers["x-chroma-token"] = @api_key
          builder.headers["x-chroma-embedding-model"] = @model
          builder.headers["Content-Type"] = "application/json"
        end
      end

      def call(texts)
        return [] if texts.nil? || texts.empty?

        payload = {
          "texts" => texts,
          "task" => "",
          "target" => "",
          "fetch_tokens" => @include_tokens ? "true" : "false"
        }

        response = @connection.post("/embed_sparse") do |req|
          req.body = JSON.generate(payload)
        end

        parse_response(response)
      end

      def embed_query(texts)
        call(texts)
      end

      def name
        NAME
      end

      def get_config
        {
          "api_key_env_var" => @api_key_env_var,
          "model" => @model,
          "include_tokens" => @include_tokens
        }
      end

      def validate_config(config)
        EmbeddingFunctions.validate_config_schema(config, "chroma-cloud-splade")
      end

      def validate_config_update(old_config, new_config)
        %w[include_tokens model].each do |key|
          next unless new_config.key?(key)
          raise ArgumentError, "Updating '#{key}' is not supported for chroma-cloud-splade" if new_config[key] != old_config[key]
        end
      end

      def self.build_from_config(config, client: nil)
        api_key_env_var = config["api_key_env_var"] || config[:api_key_env_var]
        model = config["model"] || config[:model]
        include_tokens = config.key?("include_tokens") ? config["include_tokens"] : config[:include_tokens]

        raise ArgumentError, "model must be provided in config" if model.nil?
        raise ArgumentError, "api_key_env_var must be provided in config" if api_key_env_var.nil? || api_key_env_var.to_s.empty?

        ChromaCloudSpladeEmbeddingFunction.new(
          api_key_env_var: api_key_env_var,
          model: model,
          include_tokens: include_tokens || false,
        )
      end

      def self.validate_config(config)
        EmbeddingFunctions.validate_config_schema(config, "chroma-cloud-splade")
      end

      private

      def parse_response(response)
        unless response.success?
          raise RuntimeError, "Failed to get embeddings from Chroma Cloud API: HTTP #{response.status} - #{response.body}"
        end
        data = JSON.parse(response.body)
        raw_embeddings = data["embeddings"] || []

        raw_embeddings.map do |embedding|
          if embedding.is_a?(Hash)
            indices = embedding["indices"] || []
            values = embedding["values"] || []
            labels = @include_tokens ? (embedding["labels"] || embedding["tokens"]) : nil
            Types::Validation.normalize_sparse_vector(indices: indices, values: values, labels: labels)
          elsif embedding.is_a?(Types::SparseVector)
            Types::Validation.normalize_sparse_vector(
              indices: embedding.indices,
              values: embedding.values,
              labels: @include_tokens ? embedding.labels : nil,
            )
          else
            raise ArgumentError, "Unexpected sparse embedding format: #{embedding.inspect}"
          end
        end
      rescue JSON::ParserError
        raise RuntimeError, "Invalid JSON response from Chroma Cloud API"
      end
    end

    register_sparse_embedding_function(ChromaCloudSpladeEmbeddingFunction::NAME, ChromaCloudSpladeEmbeddingFunction)
  end
end

File: clients/go/pkg/embeddings/chromacloudsplade/chromacloudsplade.go
package chromacloudsplade

import (
	"bytes"
	"context"
	"encoding/json"
	"net/http"
	"net/url"
	"strings"
	"time"

	"github.com/pkg/errors"

	chttp "github.com/chroma-core/chroma/clients/go/pkg/commons/http"
	"github.com/chroma-core/chroma/clients/go/pkg/embeddings"
)

const (
	defaultBaseURL = "https://embed.trychroma.com/embed_sparse"
	defaultModel   = "prithivida/Splade_PP_en_v1"
	defaultTimeout = 60 * time.Second
	APIKeyEnvVar   = "CHROMA_API_KEY"
)

type Client struct {
	BaseURL    string
	APIKey     embeddings.Secret `json:"-"`
	Model      embeddings.EmbeddingModel
	HTTPClient *http.Client
	Insecure   bool
}

type embeddingRequest struct {
	Texts  []string `json:"texts"`
	Task   string   `json:"task"`
	Target string   `json:"target"`
}

type sparseEmbedding struct {
	Indices []int     `json:"indices"`
	Values  []float32 `json:"values"`
}

type embeddingResponse struct {
	Embeddings []sparseEmbedding `json:"embeddings,omitempty"`
	Error      string            `json:"error,omitempty"`
}

func applyDefaults(c *Client) {
	if c.HTTPClient == nil {
		c.HTTPClient = &http.Client{Timeout: defaultTimeout}
	}
	if c.BaseURL == "" {
		c.BaseURL = defaultBaseURL
	}
	if c.Model == "" {
		c.Model = defaultModel
	}
}

func validate(c *Client) error {
	if c.APIKey.IsEmpty() {
		return errors.New("API key is required")
	}
	parsed, err := url.Parse(c.BaseURL)
	if err != nil {
		return errors.Wrap(err, "invalid base URL")
	}
	if !c.Insecure && !strings.EqualFold(parsed.Scheme, "https") {
		return errors.New("base URL must use HTTPS scheme for secure API key transmission; use WithInsecure() to override")
	}
	return nil
}

func NewClient(opts ...Option) (*Client, error) {
	client := &Client{}
	for _, opt := range opts {
		if err := opt(client); err != nil {
			return nil, errors.Wrap(err, "failed to apply option")
		}
	}
	applyDefaults(client)
	if err := validate(client); err != nil {
		return nil, errors.Wrap(err, "failed to validate client")
	}
	return client, nil
}

func (c *Client) embed(ctx context.Context, texts []string) ([]*embeddings.SparseVector, error) {
	if len(texts) == 0 {
		return make([]*embeddings.SparseVector, 0), nil
	}

	reqBody := embeddingRequest{
		Texts:  texts,
		Task:   "",
		Target: "",
	}
	reqData, err := json.Marshal(reqBody)
	if err != nil {
		return nil, errors.Wrap(err, "failed to marshal request")
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.BaseURL, bytes.NewReader(reqData))
	if err != nil {
		return nil, errors.Wrap(err, "failed to create request")
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Accept", "application/json")
	req.Header.Set("User-Agent", chttp.ChromaGoClientUserAgent)
	req.Header.Set("Cache-Control", "no-store")
	req.Header.Set("x-chroma-token", c.APIKey.Value())
	req.Header.Set("x-chroma-embedding-model", string(c.Model))

	resp, err := c.HTTPClient.Do(req)
	if err != nil {
		return nil, errors.Wrap(err, "failed to send request")
	}
	defer resp.Body.Close()

	body, err := chttp.ReadLimitedBody(resp.Body)
	if err != nil {
		return nil, errors.Wrap(err, "failed to read response")
	}

	if resp.StatusCode != http.StatusOK {
		return nil, errors.Errorf("request failed with status %d: %s", resp.StatusCode, string(body))
	}

	var embResp embeddingResponse
	if err := json.Unmarshal(body, &embResp); err != nil {
		return nil, errors.Wrap(err, "failed to unmarshal response")
	}

	if embResp.Error != "" {
		return nil, errors.Errorf("API error [status %d]: %s", resp.StatusCode, embResp.Error)
	}

	result := make([]*embeddings.SparseVector, len(embResp.Embeddings))
	for i, emb := range embResp.Embeddings {
		sv, err := embeddings.NewSparseVector(emb.Indices, emb.Values)
		if err != nil {
			return nil, errors.Wrapf(err, "failed to create sparse vector at index %d", i)
		}
		result[i] = sv
	}

	return result, nil
}

var _ embeddings.SparseEmbeddingFunction = (*EmbeddingFunction)(nil)

type EmbeddingFunction struct {
	client *Client
}

func NewEmbeddingFunction(opts ...Option) (*EmbeddingFunction, error) {
	client, err := NewClient(opts...)
	if err != nil {
		return nil, errors.Wrap(err, "failed to create client")
	}
	return &EmbeddingFunction{client: client}, nil
}

func (e *EmbeddingFunction) EmbedDocumentsSparse(ctx context.Context, documents []string) ([]*embeddings.SparseVector, error) {
	if len(documents) == 0 {
		return make([]*embeddings.SparseVector, 0), nil
	}

	vectors, err := e.client.embed(ctx, documents)
	if err != nil {
		return nil, errors.Wrap(err, "failed to embed documents")
	}

	return vectors, nil
}

func (e *EmbeddingFunction) EmbedQuerySparse(ctx context.Context, query string) (*embeddings.SparseVector, error) {
	vectors, err := e.client.embed(ctx, []string{query})
	if err != nil {
		return nil, errors.Wrap(err, "failed to embed query")
	}
	if len(vectors) == 0 {
		return nil, errors.New("no embedding returned")
	}
	return vectors[0], nil
}

func (e *EmbeddingFunction) Name() string {
	return "chroma-cloud-splade"
}

func (e *EmbeddingFunction) GetConfig() embeddings.EmbeddingFunctionConfig {
	cfg := embeddings.EmbeddingFunctionConfig{
		"model":           string(e.client.Model),
		"api_key_env_var": APIKeyEnvVar,
	}
	if e.client.Insecure {
		cfg["insecure"] = true
	}
	if e.client.BaseURL != "" {
		cfg["base_url"] = e.client.BaseURL
	}
	return cfg
}

// NewEmbeddingFunctionFromConfig creates a ChromaCloud Splade embedding function from a config map.
// Uses schema-compliant field names: api_key_env_var, model, base_url, insecure.
func NewEmbeddingFunctionFromConfig(cfg embeddings.EmbeddingFunctionConfig) (*EmbeddingFunction, error) {
	envVar, ok := cfg["api_key_env_var"].(string)
	if !ok || envVar == "" {
		return nil, errors.New("api_key_env_var is required in config")
	}
	opts := []Option{WithAPIKeyFromEnvVar(envVar)}
	if model, ok := cfg["model"].(string); ok && model != "" {
		opts = append(opts, WithModel(embeddings.EmbeddingModel(model)))
	}
	if baseURL, ok := cfg["base_url"].(string); ok && baseURL != "" {
		opts = append(opts, WithBaseURL(baseURL))
	}
	if insecure, ok := cfg["insecure"].(bool); ok && insecure {
		opts = append(opts, WithInsecure())
	} else if embeddings.AllowInsecureFromEnv() {
		embeddings.LogInsecureEnvVarWarning("ChromaCloudSplade")
		opts = append(opts, WithInsecure())
	}
	return NewEmbeddingFunction(opts...)
}

func init() {
	if err := embeddings.RegisterSparse("chroma-cloud-splade", func(cfg embeddings.EmbeddingFunctionConfig) (embeddings.SparseEmbeddingFunction, error) {
		return NewEmbeddingFunctionFromConfig(cfg)
	}); err != nil {
		panic(err)
	}
}
